apps/insights/src/components/ConformanceReport/use-download-report-for-publisher.tsx

```diff
@@ -0,0 +1,228 @@
+import { stringify as stringifyCsv } from "csv-stringify/sync";
+import {
+  addDays,
+  differenceInDays,
+  format,
+  isBefore,
+  startOfMonth,
+  startOfWeek,
+  subMonths,
+} from "date-fns";
+import { useCallback } from "react";
+import { parse } from "superjson";
+import { z } from "zod";
+
+import { WEB_API_BASE_URL } from "./constants";
+import type { Interval } from "./types";
+import { useDownloadBlob } from "../../hooks/use-download-blob";
+import { priceFeedsSchema } from "../../schemas/pyth";
+import { CLUSTER_NAMES } from "../../services/pyth";
+
+// If interval is 'daily', set interval_days=1
+// If interval is 'weekly', get the previous Sunday and set interval_days=7
+// If interval is 'monthly', get the 15th of the current month and set interval_day to the
+// difference between the 15th of the current month and the 15th of the previous month which is 28-31 days.
+const getRankingDateAndIntervalDays = (date: Date, interval: Interval) => {
+  switch (interval) {
+    case "24H": {
+      return {
+        date,
+        intervalDays: 1,
+      };
+    }
+    case "48H": {
+      return {
+        date,
+        intervalDays: 2,
+      };
+    }
+    case "72H": {
+      return {
+        date,
+        intervalDays: 3,
+      };
+    }
+    case "1W": {
+      return {
+        date: startOfWeek(date),
+        intervalDays: 7,
+      };
+    }
+    case "1M": {
+      const monthStart = startOfMonth(date);
+      let midMonth = addDays(monthStart, 14);
+      if (isBefore(date, midMonth)) {
+        midMonth = subMonths(midMonth, 1);
+      }
+      const midMonthBefore = subMonths(midMonth, 1);
+      return {
+        date: midMonth,
+        intervalDays: differenceInDays(midMonth, midMonthBefore),
+      };
+    }
+  }
+};
+
+const getFeeds = async (cluster: (typeof CLUSTER_NAMES)[number]) => {
+  const url = new URL(`/api/pyth/get-feeds`, globalThis.window.origin);
+  url.searchParams.set("cluster", cluster);
+  const data = await fetch(url);
+  const rawData = await data.text();
+  const parsedData = parse(rawData);
+  return priceFeedsSchema.element.array().parse(parsedData);
+};
+
+const PublisherQualityScoreSchema = z.object({
+  symbol: z.string(),
+  uptime_score: z.string(),
+  deviation_penalty: z.string(),
+  deviation_score: z.string(),
+  stalled_penalty: z.string(),
+  stalled_score: z.string(),
+  final_score: z.string(),
+});
+
+const PublisherQuantityScoreSchema = z.object({
+  numSymbols: z.number(),
+  rank: z.number(),
+  symbols: z.array(z.string()),
+  timestamp: z.string(),
+});
+
+const fetchRankingData = async (
+  cluster: (typeof CLUSTER_NAMES)[number],
+  publisher: string,
+  interval: Interval,
+) => {
+  const { date, intervalDays } = getRankingDateAndIntervalDays(
+    new Date(),
+    interval,
+  );
+  const quantityRankUrl = new URL(
+    `/publisher_ranking?publisher=${publisher}&cluster=${cluster}`,
```

Comment:
We should use `searchParams` instead of string interpolation just in case something ever changes with one of these and causes subtle escaping issues